File: litedb-java/src/main/java/com/litedb/memtable/MemTable.java
package com.litedb.memtable;

import java.util.*;
import java.util.concurrent.locks.ReentrantReadWriteLock;

/**
 * MemTable — In-Memory Write Buffer
 *
 * CONCEPT:
 *   The MemTable is an in-memory sorted data structure that buffers
 *   recent writes before they are flushed to disk as an SSTable.
 *
 *   Read path:
 *     1. Check MemTable first (most recent writes are here)
 *     2. If not found, check SSTables on disk (older data)
 *
 *   Write path:
 *     1. Append to WAL (durability)
 *     2. Insert into MemTable (fast in-memory write)
 *     3. When MemTable exceeds size threshold → flush to SSTable
 *
 *   This is exactly how LevelDB, RocksDB, and Cassandra work.
 *
 *   Data structure: TreeMap (Red-Black Tree) — O(log n) insert,
 *   maintains sorted order for efficient range scans and flush.
 */
public class MemTable {

    public static final String TOMBSTONE = "__DELETED__";

    // TreeMap keeps keys sorted — critical for SSTable flush and range scans
    private final TreeMap<String, String> data = new TreeMap<>();
    private final ReentrantReadWriteLock  rwLock = new ReentrantReadWriteLock();
    private final ReentrantReadWriteLock.ReadLock  readLock  = rwLock.readLock();
    private final ReentrantReadWriteLock.WriteLock writeLock = rwLock.writeLock();

    private long sizeLimitBytes;
    private long sizeBytes = 0;
    private int  writeCount = 0;

    /** Default 4 MB size limit (same as Python implementation). */
    public MemTable() {
        this(4L * 1024 * 1024);
    }

    public MemTable(long sizeLimitBytes) {
        this.sizeLimitBytes = sizeLimitBytes;
    }

    // ------------------------------------------------------------------ //
    //  Write operations                                                   //
    // ------------------------------------------------------------------ //

    /** Insert or update a key-value pair. */
    public void set(String key, String value) {
        writeLock.lock();
        try {
            if (data.containsKey(key)) {
                sizeBytes -= key.length() + data.get(key).length();
            }
            data.put(key, value);
            sizeBytes += key.length() + value.length();
            writeCount++;
        } finally {
            writeLock.unlock();
        }
    }

    /**
     * Mark a key as deleted by writing a tombstone.
     * We do NOT remove the key — the tombstone must propagate to SSTables
     * during compaction to shadow older values.
     */
    public void delete(String key) {
        writeLock.lock();
        try {
            if (data.containsKey(key)) {
                sizeBytes -= key.length() + data.get(key).length();
            }
            data.put(key, TOMBSTONE);
            sizeBytes += key.length() + TOMBSTONE.length();
            writeCount++;
        } finally {
            writeLock.unlock();
        }
    }

    // ------------------------------------------------------------------ //
    //  Read operations                                                    //
    // ------------------------------------------------------------------ //

    /**
     * Look up a key.
     * Returns:
     *   - The value string if key exists and is not deleted
     *   - TOMBSTONE if the key was deleted (caller must handle this)
     *   - null if key was never written to this MemTable
     */
    public String get(String key) {
        readLock.lock();
        try {
            return data.get(key);
        } finally {
            readLock.unlock();
        }
    }

    /**
     * Range scan: return all (key, value) pairs where startKey <= key <= endKey.
     * Returns in sorted key order (TreeMap guarantees this).
     * Includes tombstones — caller decides whether to skip them.
     */
    public List<Map.Entry<String, String>> scan(String startKey, String endKey) {
        readLock.lock();
        try {
            NavigableMap<String, String> sub = data.subMap(startKey, true, endKey, true);
            return new ArrayList<>(sub.entrySet());
        } finally {
            readLock.unlock();
        }
    }

    /**
     * Return all (key, value) pairs sorted by key.
     * Used when flushing to SSTable.
     */
    public List<Map.Entry<String, String>> itemsSorted() {
        readLock.lock();
        try {
            return new ArrayList<>(data.entrySet());
        } finally {
            readLock.unlock();
        }
    }

    // ------------------------------------------------------------------ //
    //  Size / flush threshold                                             //
    // ------------------------------------------------------------------ //

    /** Returns true when the MemTable has grown past the size limit. */
    public boolean shouldFlush() {
        return sizeBytes >= sizeLimitBytes;
    }

    public long sizeBytes() { return sizeBytes; }

    public int entryCount() {
        readLock.lock();
        try { return data.size(); } finally { readLock.unlock(); }
    }

    /**
     * Clear the MemTable after a successful flush to SSTable.
     * Called ONLY after the SSTable is safely written to disk.
     */
    public void clear() {
        writeLock.lock();
        try {
            data.clear();
            sizeBytes  = 0;
            writeCount = 0;
        } finally {
            writeLock.unlock();
        }
    }

    /** For testing: temporarily lower the limit to force a flush. */
    public void setSizeLimitBytes(long limit) {
        this.sizeLimitBytes = limit;
    }

    @Override
    public String toString() {
        return String.format("MemTable(entries=%d, size=%.1fKB, limit=%dMB)",
                entryCount(), sizeBytes / 1024.0, sizeLimitBytes / 1024 / 1024);
    }

    // ======================================================================= //
    //  DEMO                                                                    //
    // ======================================================================= //

    public static void main(String[] args) {
        System.out.println("============================================================");
        System.out.println("MEMTABLE DEMO");
        System.out.println("============================================================");

        MemTable mt = new MemTable(1024); // tiny 1KB limit for demo

        // Writes
        System.out.println("\n[Step 1] Writing key-value pairs...");
        mt.set("apple",      "red fruit");
        mt.set("banana",     "yellow fruit");
        mt.set("cherry",     "red berry");
        mt.set("date",       "brown fruit");
        mt.set("elderberry", "dark berry");
        System.out.println("  " + mt);

        // Reads
        System.out.println("\n[Step 2] Reading keys...");
        System.out.println("  GET apple    → " + mt.get("apple"));
        System.out.println("  GET banana   → " + mt.get("banana"));
        System.out.println("  GET mango    → " + mt.get("mango") + " (not found)");

        // Delete (tombstone)
        System.out.println("\n[Step 3] Deleting 'banana'...");
        mt.delete("banana");
        System.out.println("  GET banana   → '" + mt.get("banana") + "'  ← tombstone, key is deleted");

        // Range scan
        System.out.println("\n[Step 4] Range scan: 'b' to 'd'...");
        for (Map.Entry<String, String> e : mt.scan("b", "d")) {
            String marker = TOMBSTONE.equals(e.getValue()) ? " ← TOMBSTONE (deleted)" : "";
            System.out.println("  '" + e.getKey() + "': '" + e.getValue() + "'" + marker);
        }

        // Flush simulation
        System.out.println("\n[Step 5] Items sorted (as they'd be written to SSTable):");
        for (Map.Entry<String, String> e : mt.itemsSorted()) {
            System.out.println("  '" + e.getKey() + "': '" + e.getValue() + "'");
        }

        // Size threshold
        System.out.println("\n[Step 6] Should flush? " + mt.shouldFlush()
                + " (size=" + mt.sizeBytes() + " bytes, limit=1024)");

        System.out.println("\n[Done] MemTable demo complete.");
        System.out.println("\nKey insights:");
        System.out.println("  1. TreeMap keeps keys sorted — O(log n) insert");
        System.out.println("  2. Deletes write a TOMBSTONE, not an actual removal");
        System.out.println("  3. Sorted iteration is free (TreeMap)");
        System.out.println("  4. When size exceeds limit → flush to SSTable on disk");
    }
}

File: litedb-python/memtable.py
"""
memtable.py — MemTable (In-Memory Write Buffer)

CONCEPT:
  The MemTable is an in-memory sorted data structure that buffers
  recent writes before they are flushed to disk as an SSTable.

  Why sorted? Because when we flush to disk, we write keys in sorted
  order — this makes range scans and merges efficient (merge-sort).

  Read path:
    1. Check MemTable first (most recent writes are here)
    2. If not found, check SSTables on disk (older data)

  Write path:
    1. Append to WAL (durability)
    2. Insert into MemTable (fast in-memory write)
    3. When MemTable exceeds size threshold → flush to SSTable

  This is exactly how LevelDB, RocksDB, and Cassandra work.

  Data structure choice:
    We use Python's SortedDict (or a plain dict + sorted iteration).
    Production databases use a Skip List or Red-Black Tree for O(log n)
    inserts while maintaining sorted order.
    We'll use a plain dict + sort-on-flush for simplicity.
"""

import threading
from typing import Iterator

# Tombstone marker — represents a deleted key
TOMBSTONE = "__DELETED__"


class MemTable:
    """
    In-memory sorted write buffer.

    Stores the most recent value for each key.
    Deleted keys are stored with value=TOMBSTONE (not actually removed).
    This is critical: we must propagate deletes to SSTables during compaction.

    Thread-safe via a read-write lock pattern (RLock for simplicity here).
    """

    def __init__(self, size_limit_bytes: int = 4 * 1024 * 1024):  # 4MB default
        """
        size_limit_bytes: flush to SSTable when approximate size exceeds this.
        Production default: 64MB (RocksDB), 4MB (LevelDB).
        We use 4MB for demo purposes.
        """
        self._data: dict[str, str] = {}   # key → value (or TOMBSTONE)
        self._size_bytes: int = 0
        self._size_limit = size_limit_bytes
        self._lock = threading.RLock()
        self._write_count = 0

    # ------------------------------------------------------------------ #
    #  Write operations                                                    #
    # ------------------------------------------------------------------ #

    def set(self, key: str, value: str) -> None:
        """Insert or update a key-value pair."""
        with self._lock:
            # Subtract old size if key exists
            if key in self._data:
                self._size_bytes -= len(key) + len(self._data[key])

            self._data[key] = value
            self._size_bytes += len(key) + len(value)
            self._write_count += 1

    def delete(self, key: str) -> None:
        """
        Mark a key as deleted by writing a tombstone.
        We do NOT remove the key from the dict — the tombstone must
        propagate to SSTables during compaction to shadow older values.
        """
        with self._lock:
            if key in self._data:
                self._size_bytes -= len(key) + len(self._data[key])

            self._data[key] = TOMBSTONE
            self._size_bytes += len(key) + len(TOMBSTONE)
            self._write_count += 1

    # ------------------------------------------------------------------ #
    #  Read operations                                                     #
    # ------------------------------------------------------------------ #

    def get(self, key: str) -> str | None:
        """
        Look up a key.
        Returns:
          - The value string if key exists and is not deleted
          - TOMBSTONE if the key was deleted (caller must handle this)
          - None if key was never written to this MemTable
        """
        with self._lock:
            return self._data.get(key)  # None if not present

    def scan(self, start_key: str, end_key: str) -> Iterator[tuple[str, str]]:
        """
        Range scan: yield all (key, value) pairs where start_key <= key <= end_key.
        Yields in sorted key order.
        Includes tombstones — caller decides whether to skip them.
        """
        with self._lock:
            # Sort keys on the fly (production uses a skip list to avoid this)
            for key in sorted(self._data.keys()):
                if key < start_key:
                    continue
                if key > end_key:
                    break
                yield key, self._data[key]

    def items_sorted(self) -> list[tuple[str, str]]:
        """
        Return all (key, value) pairs sorted by key.
        Used when flushing to SSTable.
        """
        with self._lock:
            return sorted(self._data.items())

    # ------------------------------------------------------------------ #
    #  Size / flush threshold                                              #
    # ------------------------------------------------------------------ #

    def should_flush(self) -> bool:
        """Returns True when the MemTable has grown past the size limit."""
        return self._size_bytes >= self._size_limit

    def size_bytes(self) -> int:
        return self._size_bytes

    def entry_count(self) -> int:
        with self._lock:
            return len(self._data)

    def clear(self) -> None:
        """
        Clear the MemTable after a successful flush to SSTable.
        Called ONLY after the SSTable is safely written to disk.
        """
        with self._lock:
            self._data.clear()
            self._size_bytes = 0
            self._write_count = 0

    def __len__(self) -> int:
        return self.entry_count()

    def __repr__(self) -> str:
        return (
            f"MemTable(entries={self.entry_count()}, "
            f"size={self._size_bytes / 1024:.1f}KB, "
            f"limit={self._size_limit / 1024 / 1024:.0f}MB)"
        )


# ======================================================================= #
#  DEMO                                                                    #
# ======================================================================= #

if __name__ == "__main__":
    print("=" * 60)
    print("MEMTABLE DEMO")
    print("=" * 60)

    mt = MemTable(size_limit_bytes=1024)  # tiny 1KB limit for demo

    # --- Writes ---
    print("\n[Step 1] Writing key-value pairs...")
    mt.set("apple", "red fruit")
    mt.set("banana", "yellow fruit")
    mt.set("cherry", "red berry")
    mt.set("date", "brown fruit")
    mt.set("elderberry", "dark berry")
    print(f"  {mt}")

    # --- Read ---
    print("\n[Step 2] Reading keys...")
    print(f"  GET apple    → {mt.get('apple')}")
    print(f"  GET banana   → {mt.get('banana')}")
    print(f"  GET mango    → {mt.get('mango')} (not found)")

    # --- Delete (tombstone) ---
    print("\n[Step 3] Deleting 'banana'...")
    mt.delete("banana")
    result = mt.get("banana")
    print(f"  GET banana   → {result!r}  ← tombstone, key is deleted")

    # --- Range scan ---
    print("\n[Step 4] Range scan: 'b' to 'd'...")
    for key, value in mt.scan("b", "d"):
        marker = " ← TOMBSTONE (deleted)" if value == TOMBSTONE else ""
        print(f"  {key!r}: {value!r}{marker}")

    # --- Flush simulation ---
    print("\n[Step 5] Items sorted (as they'd be written to SSTable):")
    for key, value in mt.items_sorted():
        print(f"  {key!r}: {value!r}")

    # --- Size threshold ---
    print(f"\n[Step 6] Should flush? {mt.should_flush()} (size={mt.size_bytes()} bytes, limit=1024)")

    print("\n[Done] MemTable demo complete.")
    print("\nKey insights:")
    print("  1. All writes go into a dict — O(1) insert")
    print("  2. Deletes write a TOMBSTONE, not an actual removal")
    print("  3. Sorted iteration happens at flush time (or scan time)")
    print("  4. When size exceeds limit → flush to SSTable on disk")
